src/rust/cryptography-x509-validation/src/trust_store.rs

```diff
@@ -0,0 +1,116 @@
+// This file is dual licensed under the terms of the Apache License, Version
+// 2.0, and the BSD License. See the LICENSE file in the root of this repository
+// for complete details.
+
+use std::vec;
+
+use crate::{
+    certificate::{CertificateError, CertificateExt},
+    extensions::{ExtensionError, ExtensionExt, KeyUsage},
+};
+use cryptography_x509::{certificate::Certificate, extensions::BasicConstraints};
+
+/// Errors that can occur while constructing a `Store`.
+pub enum StoreError {
+    /// The store would be constructed in an empty state.
+    Empty,
+
+    /// The certificate is well-formed, but isn't acceptable in a store.
+    InvalidMember(&'static str),
+
+    /// The certificate isn't well-formed.
+    InvalidCert(CertificateError),
+}
+
+impl From<CertificateError> for StoreError {
+    fn from(err: CertificateError) -> Self {
+        Self::InvalidCert(err)
+    }
+}
+
+impl From<ExtensionError> for StoreError {
+    fn from(err: ExtensionError) -> Self {
+        CertificateError::InvalidExtension(err).into()
+    }
+}
+
+/// A `Store` represents a root of trust in a set of one or more CA
+/// certificates.
+///
+/// Only CA certificates (intermediates and roots) may be added to a `Store`,
+/// and only non-empty stores (i.e., with at least one CA certificate) can
+/// be constructed.
+pub struct Store<'a> {
+    trusted_certs: Vec<Certificate<'a>>,
+}
+
+impl<'a> Store<'a> {
+    fn new(certs: impl AsRef<[Certificate<'a>]>) -> Result<Self, StoreError> {
+        let mut store = Store {
+            trusted_certs: vec![],
+        };
+
+        for cert in certs.as_ref().iter() {
+            store.add(cert.clone())?;
+        }
+
+        Ok(store)
+    }
+
+    fn add(&mut self, cert: Certificate<'a>) -> Result<(), StoreError> {
+        // RFC 5280 4.2.1.9:
+        // CA certificates MUST have a BasicConstraints, it MUST be critical,
+        // and it MUST have `cA` asserted.
+        if let Some(basic_constraints) = cert.basic_constraints()? {
+            if !basic_constraints.critical {
+                return Err(StoreError::InvalidMember(
+                    "BasicConstraints must be marked critical in a CA certificate",
+                ));
+            }
+
+            let basic_constraints: BasicConstraints = basic_constraints.value()?;
+            if !basic_constraints.ca {
+                return Err(StoreError::InvalidMember(
+                    "BasicConstraints.cA must be asserted in a CA certificate",
+                ));
+            }
+        } else {
+            return Err(StoreError::InvalidMember(
+                "store certificates must have a BasicConstraints extension",
+            ));
+        }
+
+        // RFC 5280 4.2.1.3:
+        // CA certificates MUST have a KeyUsage, it SHOULD be critical,
+        // and it MUST have `keyCertSign` asserted.
+        if let Some(key_usage) = cert.key_usage()? {
+            // TODO: Check `key_usage.critical` on a policy basis here?
```

Comment:
Flagging for review: the critical bit on `KeyUsage` is optional per RFC 5280, but I wasn't sure if it made sense to allow user to opt into a stricter enforcement here.

Reply:
Definitely not for the MVP.